File: src/settings_window.py
"""
Settings window for configuring timer and audio settings.
"""

import tkinter as tk
from .config import AppConfig

class SettingsWindow:
    """Settings window for timer and audio configuration."""
    
    def __init__(self, parent, timer, audio_manager, on_settings_saved=None):
        """
        Initialize the settings window.
        
        Args:
            parent: Parent window
            timer: PomodoroTimer instance
            audio_manager: AudioManager instance
            on_settings_saved: Callback function called when settings are saved
        """
        self.timer = timer
        self.audio_manager = audio_manager
        self.on_settings_saved = on_settings_saved
        
        # Create settings window
        self.window = tk.Toplevel(parent)
        self.window.title("Settings")
        self.window.geometry(f"{AppConfig.SETTINGS_WINDOW_WIDTH}x{AppConfig.SETTINGS_WINDOW_HEIGHT}")
        self.window.configure(bg=AppConfig.COLORS['settings_bg'])
        self.window.resizable(False, False)
        
        # Center the window
        self.window.transient(parent)
        self.window.grab_set()
        
        self._create_widgets()
    
    def _create_widgets(self):
        """Create all widgets for the settings window."""
        # Title
        title_label = tk.Label(
            self.window, 
            text="Timer Settings",
            font=("Helvetica", 18, "bold"),
            bg=AppConfig.COLORS['settings_bg'],
            fg=AppConfig.COLORS['text_primary']
        )
        title_label.pack(pady=20)
        
        # Settings frame
        settings_frame = tk.Frame(self.window, bg=AppConfig.COLORS['settings_bg'])
        settings_frame.pack(pady=10)
        
        # Timer settings
        self._create_timer_settings(settings_frame)
        
        # Volume settings
        self._create_volume_settings(settings_frame)
        
        # Buttons
        self._create_buttons()
    
    def _create_timer_settings(self, parent):
        """Create timer configuration widgets."""
        # Work time setting
        work_frame = tk.Frame(parent, bg=AppConfig.COLORS['settings_bg'])
        work_frame.pack(pady=10)
        tk.Label(
            work_frame,
            text="Work Time (minutes):",
            font=("Helvetica", 12),
            bg=AppConfig.COLORS['settings_bg'],
            fg=AppConfig.COLORS['text_primary']
        ).pack()
        
        self.work_time_var = tk.StringVar(value=str(self.timer.get_work_time_minutes()))
        self.work_time_entry = tk.Entry(
            work_frame,
            textvariable=self.work_time_var,
            font=("Helvetica", 12),
            width=10,
            justify="center",
            relief="flat",
            bg=AppConfig.COLORS['text_primary'],
            fg=AppConfig.COLORS['background']
        )
        self.work_time_entry.pack(pady=5)
        
        # Short break time setting
        short_frame = tk.Frame(parent, bg=AppConfig.COLORS['settings_bg'])
        short_frame.pack(pady=10)
        tk.Label(
            short_frame,
            text="Short Break Time (minutes):",
            font=("Helvetica", 12),
            bg=AppConfig.COLORS['settings_bg'],
            fg=AppConfig.COLORS['text_primary']
        ).pack()
        
        self.short_break_var = tk.StringVar(value=str(self.timer.get_short_break_minutes()))
        self.short_break_entry = tk.Entry(
            short_frame,
            textvariable=self.short_break_var,
            font=("Helvetica", 12),
            width=10,
            justify="center",
            relief="flat",
            bg=AppConfig.COLORS['text_primary'],
            fg=AppConfig.COLORS['background']
        )
        self.short_break_entry.pack(pady=5)
        
        # Long break time setting
        long_frame = tk.Frame(parent, bg=AppConfig.COLORS['settings_bg'])
        long_frame.pack(pady=10)
        tk.Label(
            long_frame,
            text="Long Break Time (minutes):",
            font=("Helvetica", 12),
            bg=AppConfig.COLORS['settings_bg'],
            fg=AppConfig.COLORS['text_primary']
        ).pack()
        
        self.long_break_var = tk.StringVar(value=str(self.timer.get_long_break_minutes()))
        self.long_break_entry = tk.Entry(
            long_frame,
            textvariable=self.long_break_var,
            font=("Helvetica", 12),
            width=10,
            justify="center",
            relief="flat",
            bg=AppConfig.COLORS['text_primary'],
            fg=AppConfig.COLORS['background']
        )
        self.long_break_entry.pack(pady=5)
    
    def _create_volume_settings(self, parent):
        """Create volume configuration widgets."""
        volume_frame = tk.Frame(parent, bg=AppConfig.COLORS['settings_bg'])
        volume_frame.pack(pady=15)
        
        # Music volume setting
        music_vol_frame = tk.Frame(volume_frame, bg=AppConfig.COLORS['settings_bg'])
        music_vol_frame.pack(pady=5)
        tk.Label(
            music_vol_frame,
            text="Music Volume:",
            font=("Helvetica", 12),
            bg=AppConfig.COLORS['settings_bg'],
            fg=AppConfig.COLORS['text_primary']
        ).pack()
        
        self.music_volume_scale = tk.Scale(
            music_vol_frame,
            from_=0,
            to=100,
            orient=tk.HORIZONTAL,
            bg=AppConfig.COLORS['background'],
            fg=AppConfig.COLORS['text_primary'],
            highlightthickness=0,
            troughcolor=AppConfig.COLORS['settings_bg'],
            activebackground=AppConfig.COLORS['slider_active_blue'],
            length=200
        )
        self.music_volume_scale.set(int(self.audio_manager.get_music_volume() * 100))
        self.music_volume_scale.pack(pady=5)
        
        # Alarm volume setting
        alarm_vol_frame = tk.Frame(volume_frame, bg=AppConfig.COLORS['settings_bg'])
        alarm_vol_frame.pack(pady=5)
        tk.Label(
            alarm_vol_frame,
            text="Alarm Volume:",
            font=("Helvetica", 12),
            bg=AppConfig.COLORS['settings_bg'],
            fg=AppConfig.COLORS['text_primary']
        ).pack()
        
        self.alarm_volume_scale = tk.Scale(
            alarm_vol_frame,
            from_=0,
            to=100,
            orient=tk.HORIZONTAL,
            bg=AppConfig.COLORS['background'],
            fg=AppConfig.COLORS['text_primary'],
            highlightthickness=0,
            troughcolor=AppConfig.COLORS['settings_bg'],
            activebackground=AppConfig.COLORS['slider_active_red'],
            length=200
        )
        self.alarm_volume_scale.set(int(self.audio_manager.get_alarm_volume() * 100))
        self.alarm_volume_scale.pack(pady=5)
    
    def _create_buttons(self):
        """Create save and cancel buttons."""
        button_frame = tk.Frame(self.window, bg=AppConfig.COLORS['settings_bg'])
        button_frame.pack(pady=30)
        
        save_button = tk.Button(
            button_frame,
            text="Save",
            command=self._save_settings,
            font=("Helvetica", 12, "bold"),
            bg=AppConfig.COLORS['button_green'],
            fg="black",
            activebackground=AppConfig.COLORS['button_green_active'],
            relief="flat",
            padx=20,
            pady=8,
            cursor="hand2"
        )
        save_button.pack(side=tk.LEFT, padx=15)
        
        cancel_button = tk.Button(
            button_frame,
            text="Cancel",
            command=self._cancel_settings,
            font=("Helvetica", 12, "bold"),
            bg=AppConfig.COLORS['button_red'],
            fg="black",
            activebackground=AppConfig.COLORS['button_red_active'],
            relief="flat",
            padx=20,
            pady=8,
            cursor="hand2"
        )
        cancel_button.pack(side=tk.LEFT, padx=15)
    
    def _save_settings(self):
        """Save all settings and close the window."""
        try:
            # Validate and convert input
            work_mins = int(self.work_time_var.get())
            short_mins = int(self.short_break_var.get())
            long_mins = int(self.long_break_var.get())
            
            # Check for reasonable values
            if work_mins < AppConfig.MIN_WORK_TIME or work_mins > AppConfig.MAX_WORK_TIME:
                raise ValueError(f"Work time must be between {AppConfig.MIN_WORK_TIME} and {AppConfig.MAX_WORK_TIME} minutes")
            if short_mins < AppConfig.MIN_SHORT_BREAK or short_mins > AppConfig.MAX_SHORT_BREAK:
                raise ValueError(f"Short break must be between {AppConfig.MIN_SHORT_BREAK} and {AppConfig.MAX_SHORT_BREAK} minutes")
            if long_mins < AppConfig.MIN_LONG_BREAK or long_mins > AppConfig.MAX_LONG_BREAK:
                raise ValueError(f"Long break must be between {AppConfig.MIN_LONG_BREAK} and {AppConfig.MAX_LONG_BREAK} minutes")
            
            # Update timer settings
            self.timer.set_work_time(work_mins)
            self.timer.set_short_break_time(short_mins)
            self.timer.set_long_break_time(long_mins)
            
            # Update volume settings
            music_volume = self.music_volume_scale.get() / 100.0
            alarm_volume = self.alarm_volume_scale.get() / 100.0
            self.audio_manager.set_music_volume(music_volume)
            self.audio_manager.set_alarm_volume(alarm_volume)
            
            # Call callback if provided
            if self.on_settings_saved:
                self.on_settings_saved()
            
            self.window.destroy()
            
        except ValueError as e:
            self._show_error(str(e))
    
    def _cancel_settings(self):
        """Cancel settings changes and close the window."""
        self.window.destroy()
    
    def _show_error(self, message):
        """Show an error message."""
        # Clear any existing error labels
        for widget in self.window.winfo_children():
            if isinstance(widget, tk.Label) and widget.cget("fg") == AppConfig.COLORS['button_red']:
                widget.destroy()
        
        error_label = tk.Label(
            self.window,
            text=message,
            fg=AppConfig.COLORS['button_red'],
            bg=AppConfig.COLORS['settings_bg'],
            font=("Helvetica", 10)
        )
        error_label.pack(pady=5)
        self.window.after(3000, error_label.destroy)  # Remove error after 3 seconds

File: src/main/java/com/gooddaytolearn/SettingsWindow.java
package com.gooddaytolearn;

import javax.swing.*;
import javax.swing.border.EmptyBorder;
import javax.swing.filechooser.FileNameExtensionFilter;
import java.awt.*;
import java.io.File;

/**
 * Settings window for configuring timer and audio settings.
 */
public class SettingsWindow extends JDialog {
    
    private PomodoroTimer timer;
    private AudioManager audioManager;
    private Runnable onSettingsSaved;
    
    private JSpinner workTimeSpinner;
    private JSpinner shortBreakSpinner;
    private JSpinner longBreakSpinner;
    private JSlider musicVolumeSlider;
    private JSlider alarmVolumeSlider;
    
    // Custom sound file components
    private JLabel customMusicLabel;
    private JButton selectMusicButton;
    private JButton resetMusicButton;
    private JLabel customAlarmLabel;
    private JButton selectAlarmButton;
    private JButton resetAlarmButton;
    
    /**
     * Initialize the settings window.
     * 
     * @param parent Parent window
     * @param timer PomodoroTimer instance
     * @param audioManager AudioManager instance
     * @param onSettingsSaved Callback function called when settings are saved
     */
    public SettingsWindow(JFrame parent, PomodoroTimer timer, AudioManager audioManager, Runnable onSettingsSaved) {
        super(parent, "Settings", true);
        this.timer = timer;
        this.audioManager = audioManager;
        this.onSettingsSaved = onSettingsSaved;
        
        setupWindow();
        createWidgets();
        pack();
        setLocationRelativeTo(parent);
    }
    
    /**
     * Configure the settings window.
     */
    private void setupWindow() {
        setSize(AppConfig.SETTINGS_WINDOW_WIDTH, AppConfig.SETTINGS_WINDOW_HEIGHT);
        setResizable(false);
        getContentPane().setBackground(AppConfig.COLORS.get("settings_bg"));
        setDefaultCloseOperation(DISPOSE_ON_CLOSE);
    }
    
    /**
     * Create all widgets for the settings window with modern styling.
     */
    private void createWidgets() {
        setLayout(new BorderLayout());
        
        // Title with modern typography
        JLabel titleLabel = new JLabel("Timer Settings", SwingConstants.CENTER);
        titleLabel.setFont(new Font("SF Pro Display", Font.BOLD, 20));
        titleLabel.setForeground(AppConfig.COLORS.get("text_primary"));
        titleLabel.setBorder(new EmptyBorder(25, 0, 25, 0));
        add(titleLabel, BorderLayout.NORTH);
        
        // Settings panel with modern card-like appearance
        JPanel settingsPanel = new JPanel(new GridBagLayout()) {
            @Override
            protected void paintComponent(Graphics g) {
                Graphics2D g2d = (Graphics2D) g;
                g2d.setRenderingHint(RenderingHints.KEY_ANTIALIASING, RenderingHints.VALUE_ANTIALIAS_ON);
                
                // Draw rounded card background
                g2d.setColor(AppConfig.COLORS.get("panel_bg"));
                g2d.fillRoundRect(10, 10, getWidth() - 20, getHeight() - 20, 16, 16);
                
                // Add subtle border
                g2d.setColor(AppConfig.COLORS.get("border_color"));
                g2d.drawRoundRect(10, 10, getWidth() - 20, getHeight() - 20, 16, 16);
            }
        };
        settingsPanel.setOpaque(false);
        settingsPanel.setBorder(new EmptyBorder(20, 40, 20, 40));
        
        GridBagConstraints gbc = new GridBagConstraints();
        gbc.insets = new Insets(8, 8, 8, 8);
        gbc.anchor = GridBagConstraints.WEST;
        
        // Timer settings
        addTimerSettings(settingsPanel, gbc);
        
        // Audio settings
        addAudioSettings(settingsPanel, gbc);
        
        add(settingsPanel, BorderLayout.CENTER);
        
        // Buttons
        addButtons();
    }
    
    /**
     * Add timer settings controls.
     */
    private void addTimerSettings(JPanel panel, GridBagConstraints gbc) {
        // Work time
        gbc.gridx = 0; gbc.gridy = 0;
        JLabel workLabel = createLabel("Work Time (minutes):");
        panel.add(workLabel, gbc);
        
        gbc.gridx = 1;
        workTimeSpinner = new JSpinner(new SpinnerNumberModel(timer.getWorkTimeMinutes(), 
                AppConfig.MIN_WORK_TIME, AppConfig.MAX_WORK_TIME, 1));
        styleSpinner(workTimeSpinner);
        panel.add(workTimeSpinner, gbc);
        
        // Short break time
        gbc.gridx = 0; gbc.gridy = 1;
        JLabel shortBreakLabel = createLabel("Short Break (minutes):");
        panel.add(shortBreakLabel, gbc);
        
        gbc.gridx = 1;
        shortBreakSpinner = new JSpinner(new SpinnerNumberModel(timer.getShortBreakMinutes(), 1, 30, 1));
        styleSpinner(shortBreakSpinner);
        panel.add(shortBreakSpinner, gbc);
        
        // Long break time
        gbc.gridx = 0; gbc.gridy = 2;
        JLabel longBreakLabel = createLabel("Long Break (minutes):");
        panel.add(longBreakLabel, gbc);
        
        gbc.gridx = 1;
        longBreakSpinner = new JSpinner(new SpinnerNumberModel(timer.getLongBreakMinutes(), 1, 60, 1));
        styleSpinner(longBreakSpinner);
        panel.add(longBreakSpinner, gbc);
    }
    
    /**
     * Add audio settings controls.
     */
    private void addAudioSettings(JPanel panel, GridBagConstraints gbc) {
        // Custom music file section
        gbc.gridx = 0; gbc.gridy = 3;
        gbc.gridwidth = 2;
        gbc.fill = GridBagConstraints.HORIZONTAL;
        
        JLabel customMusicSectionLabel = createLabel("Custom Background Music:");
        customMusicSectionLabel.setFont(new Font("SF Pro Text", Font.BOLD, 13));
        panel.add(customMusicSectionLabel, gbc);
        
        // Music file display and controls
        gbc.gridy = 4;
        JPanel musicFilePanel = new JPanel(new FlowLayout(FlowLayout.LEFT, 0, 0));
        musicFilePanel.setOpaque(false);
        
        customMusicLabel = new JLabel(getMusicFileName());
        customMusicLabel.setForeground(AppConfig.COLORS.get("text_secondary"));
        customMusicLabel.setFont(new Font("SF Pro Text", Font.PLAIN, 12));
        musicFilePanel.add(customMusicLabel);
        
        musicFilePanel.add(Box.createHorizontalStrut(10));
        
        selectMusicButton = createSmallButton("Browse...");
        selectMusicButton.addActionListener(e -> selectCustomMusicFile());
        musicFilePanel.add(selectMusicButton);
        
        musicFilePanel.add(Box.createHorizontalStrut(5));
        
        resetMusicButton = createSmallButton("Reset");
        resetMusicButton.addActionListener(e -> resetMusicToDefault());
        musicFilePanel.add(resetMusicButton);
        
        panel.add(musicFilePanel, gbc);
        
        // Music volume
        gbc.gridy = 5;
        JLabel musicVolumeLabel = createLabel("Music Volume:");
        panel.add(musicVolumeLabel, gbc);
        
        gbc.gridy = 6;
        musicVolumeSlider = new JSlider(0, 100, (int)(audioManager.getMusicVolume() * 100));
        styleSlider(musicVolumeSlider);
        panel.add(musicVolumeSlider, gbc);
        
        // Custom alarm file section
        gbc.gridy = 7;
        JLabel customAlarmSectionLabel = createLabel("Custom Alarm Sound:");
        customAlarmSectionLabel.setFont(new Font("SF Pro Text", Font.BOLD, 13));
        panel.add(customAlarmSectionLabel, gbc);
        
        // Alarm file display and controls
        gbc.gridy = 8;
        JPanel alarmFilePanel = new JPanel(new FlowLayout(FlowLayout.LEFT, 0, 0));
        alarmFilePanel.setOpaque(false);
        
        customAlarmLabel = new JLabel(getAlarmFileName());
        customAlarmLabel.setForeground(AppConfig.COLORS.get("text_secondary"));
        customAlarmLabel.setFont(new Font("SF Pro Text", Font.PLAIN, 12));
        alarmFilePanel.add(customAlarmLabel);
        
        alarmFilePanel.add(Box.createHorizontalStrut(10));
        
        selectAlarmButton = createSmallButton("Browse...");
        selectAlarmButton.addActionListener(e -> selectCustomAlarmFile());
        alarmFilePanel.add(selectAlarmButton);
        
        alarmFilePanel.add(Box.createHorizontalStrut(5));
        
        resetAlarmButton = createSmallButton("Reset");
        resetAlarmButton.addActionListener(e -> resetAlarmToDefault());
        alarmFilePanel.add(resetAlarmButton);
        
        panel.add(alarmFilePanel, gbc);
        
        // Alarm volume
        gbc.gridy = 9;
        JLabel alarmVolumeLabel = createLabel("Alarm Volume:");
        panel.add(alarmVolumeLabel, gbc);
        
        gbc.gridy = 10;
        alarmVolumeSlider = new JSlider(0, 100, (int)(audioManager.getAlarmVolume() * 100));
        styleSlider(alarmVolumeSlider);
        panel.add(alarmVolumeSlider, gbc);
    }
    
    /**
     * Add modern action buttons.
     */
    private void addButtons() {
        JPanel buttonPanel = new JPanel(new FlowLayout(FlowLayout.CENTER, 15, 0));
        buttonPanel.setBackground(AppConfig.COLORS.get("settings_bg"));
        buttonPanel.setBorder(new EmptyBorder(20, 0, 25, 0));
        
        // Save button
        JButton saveButton = createModernButton("Save", "button_green", "button_green_active");
        saveButton.addActionListener(e -> saveSettings());
        buttonPanel.add(saveButton);
        
        // Cancel button
        JButton cancelButton = createModernButton("Cancel", "button_red", "button_red_active");
        cancelButton.addActionListener(e -> dispose());
        buttonPanel.add(cancelButton);
        
        add(buttonPanel, BorderLayout.SOUTH);
    }
    
    /**
     * Create a modern styled button for settings window.
     */
    private JButton createModernButton(String text, String bgColorKey, String activeColorKey) {
        JButton button = new JButton(text) {
            @Override
            protected void paintComponent(Graphics g) {
                Graphics2D g2d = (Graphics2D) g;
                g2d.setRenderingHint(RenderingHints.KEY_ANTIALIASING, RenderingHints.VALUE_ANTIALIAS_ON);
                
                // Draw rounded background
                Color bgColor = getModel().isPressed() || getModel().isRollover() ? 
                    AppConfig.COLORS.get(activeColorKey) : AppConfig.COLORS.get(bgColorKey);
                g2d.setColor(bgColor);
                g2d.fillRoundRect(0, 0, getWidth(), getHeight(), 10, 10);
                
                // Draw text
                FontMetrics fm = g2d.getFontMetrics();
                int textX = (getWidth() - fm.stringWidth(getText())) / 2;
                int textY = (getHeight() + fm.getAscent() - fm.getDescent()) / 2;
                
                g2d.setColor(getForeground());
                g2d.drawString(getText(), textX, textY);
            }
        };
        
        button.setFont(new Font("SF Pro Text", Font.BOLD, 13));
        button.setForeground(Color.WHITE);
        button.setFocusPainted(false);
        button.setBorderPainted(false);
        button.setContentAreaFilled(false);
        button.setPreferredSize(new Dimension(80, 35));
        button.setCursor(new Cursor(Cursor.HAND_CURSOR));
        
        return button;
    }
    
    /**
     * Create a styled label with modern typography.
     */
    private JLabel createLabel(String text) {
        JLabel label = new JLabel(text);
        label.setForeground(AppConfig.COLORS.get("text_primary"));
        label.setFont(new Font("SF Pro Text", Font.PLAIN, 13));
        return label;
    }
    
    /**
     * Style a spinner component with dark theme.
     */
    private void styleSpinner(JSpinner spinner) {
        spinner.setPreferredSize(new Dimension(90, 30));
        
        // Style the editor
        JComponent editor = spinner.getEditor();
        if (editor instanceof JSpinner.DefaultEditor) {
            JTextField textField = ((JSpinner.DefaultEditor) editor).getTextField();
            textField.setBackground(AppConfig.COLORS.get("button_bg"));
            textField.setForeground(AppConfig.COLORS.get("text_primary"));
            textField.setCaretColor(AppConfig.COLORS.get("text_primary"));
            textField.setBorder(BorderFactory.createEmptyBorder(5, 8, 5, 8));
        }
        
        // Style the spinner buttons
        for (Component comp : spinner.getComponents()) {
            if (comp instanceof JButton) {
                JButton button = (JButton) comp;
                button.setBackground(AppConfig.COLORS.get("button_bg"));
                button.setForeground(AppConfig.COLORS.get("text_primary"));
                button.setBorder(BorderFactory.createEmptyBorder());
            }
        }
    }
    
    /**
     * Style a slider component with modern dark theme.
     */
    private void styleSlider(JSlider slider) {
        slider.setOpaque(false);
        slider.setForeground(AppConfig.COLORS.get("text_primary"));
        slider.setPaintTicks(false);
        slider.setPaintLabels(false);
        slider.setMajorTickSpacing(25);
        slider.setMinorTickSpacing(5);
        
        // Set custom UI for better dark theme support
        slider.putClientProperty("Slider.trackColor", AppConfig.COLORS.get("slider_track"));
        slider.putClientProperty("Slider.thumbColor", AppConfig.COLORS.get("accent_primary"));
    }
    
    /**
     * Save the current settings.
     */
    private void saveSettings() {
        // Update timer settings
        timer.setWorkTime((Integer) workTimeSpinner.getValue());
        timer.setShortBreakTime((Integer) shortBreakSpinner.getValue());
        timer.setLongBreakTime((Integer) longBreakSpinner.getValue());
        
        // Update audio settings
        audioManager.setMusicVolume(musicVolumeSlider.getValue() / 100.0);
        audioManager.setAlarmVolume(alarmVolumeSlider.getValue() / 100.0);
        
        // Callback to parent
        if (onSettingsSaved != null) {
            onSettingsSaved.run();
        }
        
        dispose();
    }
    
    /**
     * Create a smaller button for file selection.
     */
    private JButton createSmallButton(String text) {
        JButton button = new JButton(text) {
            @Override
            protected void paintComponent(Graphics g) {
                Graphics2D g2d = (Graphics2D) g;
                g2d.setRenderingHint(RenderingHints.KEY_ANTIALIASING, RenderingHints.VALUE_ANTIALIAS_ON);
                
                // Draw rounded background
                Color bgColor = getModel().isPressed() || getModel().isRollover() ? 
                    AppConfig.COLORS.get("button_blue_active") : AppConfig.COLORS.get("button_blue");
                g2d.setColor(bgColor);
                g2d.fillRoundRect(0, 0, getWidth(), getHeight(), 6, 6);
                
                // Draw text
                FontMetrics fm = g2d.getFontMetrics();
                int textX = (getWidth() - fm.stringWidth(getText())) / 2;
                int textY = (getHeight() + fm.getAscent() - fm.getDescent()) / 2;
                
                g2d.setColor(getForeground());
                g2d.drawString(getText(), textX, textY);
            }
        };
        
        button.setFont(new Font("SF Pro Text", Font.PLAIN, 11));
        button.setForeground(Color.WHITE);
        button.setFocusPainted(false);
        button.setBorderPainted(false);
        button.setContentAreaFilled(false);
        button.setPreferredSize(new Dimension(60, 25));
        button.setCursor(new Cursor(Cursor.HAND_CURSOR));
        
        return button;
    }
    
    /**
     * Get the display name for the current music file.
     */
    private String getMusicFileName() {
        String customFile = audioManager.getCustomMusicFile();
        if (customFile != null) {
            return new File(customFile).getName();
        }
        return "Default (rain.wav)";
    }
    
    /**
     * Get the display name for the current alarm file.
     */
    private String getAlarmFileName() {
        String customFile = audioManager.getCustomAlarmFile();
        if (customFile != null) {
            return new File(customFile).getName();
        }
        return "Default (alarm.wav)";
    }
    
    /**
     * Open file chooser to select custom music file.
     */
    private void selectCustomMusicFile() {
        JFileChooser fileChooser = createAudioFileChooser();
        fileChooser.setDialogTitle("Select Background Music File");
        
        if (fileChooser.showOpenDialog(this) == JFileChooser.APPROVE_OPTION) {
            File selectedFile = fileChooser.getSelectedFile();
            audioManager.setCustomMusicFile(selectedFile.getAbsolutePath());
            customMusicLabel.setText(selectedFile.getName());
        }
    }
    
    /**
     * Open file chooser to select custom alarm file.
     */
    private void selectCustomAlarmFile() {
        JFileChooser fileChooser = createAudioFileChooser();
        fileChooser.setDialogTitle("Select Alarm Sound File");
        
        if (fileChooser.showOpenDialog(this) == JFileChooser.APPROVE_OPTION) {
            File selectedFile = fileChooser.getSelectedFile();
            audioManager.setCustomAlarmFile(selectedFile.getAbsolutePath());
            customAlarmLabel.setText(selectedFile.getName());
        }
    }
    
    /**
     * Reset music to default.
     */
    private void resetMusicToDefault() {
        audioManager.resetMusicToDefault();
        customMusicLabel.setText("Default (rain.wav)");
    }
    
    /**
     * Reset alarm to default.
     */
    private void resetAlarmToDefault() {
        audioManager.resetAlarmToDefault();
        customAlarmLabel.setText("Default (alarm.wav)");
    }
    
    /**
     * Create a file chooser configured for audio files.
     */
    private JFileChooser createAudioFileChooser() {
        JFileChooser fileChooser = new JFileChooser();
        fileChooser.setFileSelectionMode(JFileChooser.FILES_ONLY);
        
        // Add audio file filters
        FileNameExtensionFilter audioFilter = new FileNameExtensionFilter(
            "Audio Files (*.wav, *.mp3, *.aiff, *.au)", 
            "wav", "mp3", "aiff", "au"
        );
        FileNameExtensionFilter wavFilter = new FileNameExtensionFilter("WAV Files (*.wav)", "wav");
        FileNameExtensionFilter mp3Filter = new FileNameExtensionFilter("MP3 Files (*.mp3)", "mp3");
        FileNameExtensionFilter aiffFilter = new FileNameExtensionFilter("AIFF Files (*.aiff)", "aiff");
        FileNameExtensionFilter auFilter = new FileNameExtensionFilter("AU Files (*.au)", "au");
        
        fileChooser.addChoosableFileFilter(audioFilter);
        fileChooser.addChoosableFileFilter(wavFilter);
        fileChooser.addChoosableFileFilter(mp3Filter);
        fileChooser.addChoosableFileFilter(aiffFilter);
        fileChooser.addChoosableFileFilter(auFilter);
        fileChooser.setFileFilter(audioFilter);
        
        return fileChooser;
    }
}
